rmf_site_editor/src/widgets/console.rs

```diff
@@ -0,0 +1,471 @@
+/*
+ * Copyright (C) 2022 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+use crate::{site::*, widgets::AppEvents};
+use bevy::{ecs::system::SystemParam, prelude::*};
+use bevy_egui::{
+    egui::{self, CollapsingHeader, Color32, FontId, RichText, Ui},
+    EguiContext,
+};
+use bevy_utils::tracing::{field::Field, span::Record, Event, Id, Level, Subscriber};
+use crossbeam_channel::{unbounded, Receiver, SendError, Sender, TryRecvError};
+use std::collections::HashMap;
+use std::fmt::{self, Debug, Write};
+use tracing_subscriber::{field::Visit, layer::Context, prelude::*, EnvFilter, Layer};
+
+#[derive(Debug, Clone, Copy, Eq, Hash, PartialEq)]
+pub enum LogCategory {
+    Status = 0,
+    Warning = 1,
+    Error = 2,
+    Bevy = 3,
+    Hint = 4,
+}
+
+impl fmt::Display for LogCategory {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            LogCategory::Hint => write!(f, "[HINT] "),
+            LogCategory::Status => write!(f, "[STATUS] "),
+            LogCategory::Warning => write!(f, "[WARNING] "),
+            LogCategory::Error => write!(f, "[ERROR] "),
+            LogCategory::Bevy => write!(f, "[BEVY] "),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Component)]
+pub struct Log {
+    pub category: LogCategory,
+    pub message: String,
+}
+
+#[derive(Resource)]
+pub struct LogHistory {
+    log_history: Vec<Log>,
+    current_log: Option<Log>,
+    category_filter: HashMap<LogCategory, bool>,
+    checked_all: bool,        // True if "All" box is checked
+    stored_checked_all: bool, // Stored state of "All" checkbox
+    display_limit: usize,
+    show_full_history: bool,
+    category_count: Vec<usize>,
+    receiver: Receiver<Log>,
+}
+
+impl Default for LogHistory {
+    fn default() -> Self {
+        let mut filter_hashmap = HashMap::new();
+        filter_hashmap.insert(LogCategory::Status, true);
+        filter_hashmap.insert(LogCategory::Warning, true);
+        filter_hashmap.insert(LogCategory::Error, true);
+        filter_hashmap.insert(LogCategory::Bevy, false);
```

Comment:
I noticed that this filter only gets applied to the console that is expanded, not to the one that prints messages at every iteration, which means `[BEVY]` messages are still visibile by default (and spam that console a bit at the moment).
Is this intended?

Reply:
Yup, currently the bar above the expandable console shows that latest log that comes through regardless of whether its category is being filtered or not. Do you think it would be necessary to have a button that disables Bevy logs being shown in the bar? Or should they not even be shown at all?

Reply:
:man_shrugging:  Personally I found it a bit confusing that the `BEVY` level is selected to be filtered out but was still displayed in one of the two logs, I would have expected the two displays to be coherent

Reply:
[12bdc1d](https://github.com/open-rmf/rmf_site/pull/121/commits/12bdc1d07916a14bf8c099abe330b02f5d16c151)